apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/SpannerCommand.cs

```diff
@@ -0,0 +1,367 @@
+            }
+            else
+            {
+                var w = new Mutation.Types.Delete {Table = SpannerCommandTextBuilder.TargetTable};
+                w.KeySet.Keys.Add(new ListValue {
+                    Values = {Parameters.Cast<SpannerParameter>().Select(x => TypeMap.ToValue(x.Value, x.TypeCode))}
+                });
+                switch (SpannerCommandTextBuilder.SpannerCommandType)
+                {
+                    case SpannerCommandType.Delete:
+                        mutations.Add(new Mutation {
+                            Delete = w
+                        });
+                        break;
+                    default:
+                        throw new ArgumentOutOfRangeException();
+                }
+            }
+
+            // Make the request.  This will commit immediately or not depending on whether a transaction was explicitly created.
+            await GetSpannerTransaction().ExecuteMutationsAsync(mutations, cancellationToken)
+                .WithTimeout(TimeSpan.FromSeconds(CommandTimeout), "The timeout of the SpannerCommand was exceeded.");
+            // Return the number of records affected.
+            return mutations.Count;
+        }
+
+        /// <inheritdoc />
+        public override object ExecuteScalar()
+        {
+            return ExecuteScalarAsync(_synchronousCancellationTokenSource.Token).Result;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <returns></returns>
+        public Task<T> ExecuteScalarAsync<T>()
+        {
+            return ExecuteScalarAsync<T>(CancellationToken.None);
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <typeparam name="T"></typeparam>
+        /// <returns></returns>
+        public async Task<T> ExecuteScalarAsync<T>(CancellationToken cancellationToken)
+        {
+            using (var reader = await ExecuteDbDataReaderAsync(CommandBehavior.SingleRow, cancellationToken)
+                .ConfigureAwait(false))
+            {
+                if (await reader.ReadAsync(cancellationToken).ConfigureAwait(false) && reader.HasRows &&
+                    reader.FieldCount > 0)
+                    return reader.GetFieldValue<T>(0);
+            }
+            return default(T);
+        }
+
+        /// <inheritdoc />
+        public override async Task<object> ExecuteScalarAsync(CancellationToken cancellationToken)
+        {
+            using (var reader = await ExecuteDbDataReaderAsync(CommandBehavior.SingleRow, cancellationToken)
+                .ConfigureAwait(false))
+            {
+                if (await reader.ReadAsync(cancellationToken).ConfigureAwait(false) && reader.HasRows &&
+                    reader.FieldCount > 0)
+                    return reader[0];
+            }
+            return null;
+        }
+
+        /// <inheritdoc />
+        public override void Prepare()
```

Comment:
"If the CommandType property is set to TableDirect, Prepare does nothing. If CommandType is set to StoredProcedure, the call to Prepare should succeed, although it may result in a no-op."

Prepare for commands (not TransactionScope) can simply be a No-op as its a single resource and we can simply log that you are making an otherwise unnecessary call.  Also prepare in the command context usually means "make whatever optimizations you need because I'm going to call execute a lot".

It's only an exception level error if prepare is called on a the resource manager (which is when the user uses spanner in a TransactionScope with other resources).  In that case, two phase commit is required and we throw a good error.